src/viam/sdk/common/proto_value.hpp

```diff
@@ -0,0 +1,324 @@
+#pragma once
+
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace google {
+namespace protobuf {
+
+// Forward declaration of google::protobuf Value and Struct.
+// The class below is written so as to keep Value out of the ABI, and as such can be instantiated
+// with Value as an incomplete type.
+
+class Value;
+class Struct;
+}  // namespace protobuf
+}  // namespace google
+
+namespace viam {
+namespace sdk {
+
+namespace impl {
+
+struct move_may_throw {
+    move_may_throw() noexcept(false) = default;
+    move_may_throw& operator=(move_may_throw&&) noexcept(false) = default;
+};
+
+// Type trait for determining if move operations on ProtoValue are noexcept.
+// Conditional noexcept-ness of move construction on vector and unordered_map is not
+// guaranteed until c++17, and even then not fully. Thus define a traits class to check whether the
+// containers have nothrow move ops, and use that to determine if ProtoValue's move ops are nothrow.
+// We use move_may_throw as an illustrative dummy class since, perhaps surprisingly, nothrow
+// movability of vector and unordered_map doesn't depend on their stored type.
+struct all_proto_moves_noexcept
+    : std::integral_constant<bool,
+                             std::is_nothrow_move_constructible<std::vector<move_may_throw>>{} &&
+                                 std::is_nothrow_move_constructible<
+                                     std::unordered_map<std::string, move_may_throw>>{}> {};
+
+}  // namespace impl
+
+/// @brief Type-erased value for storing google::protobuf::Value types.
+/// A ProtoValue can be nullptr, bool, int, double, std::string, or, recursively, a vector or
+/// string-map of ProtoValue.
+/// This type is used at API/ABI boundaries for interfacing with grpc/proto code.
+class ProtoValue {
```

Comment:
I think it is great how much of this was able to exist out of the header.

Reply:
Same! And only one round of explicit insantiations needed, a bunch of other ones happen implicitly.

I was wondering if *all* of it could be implicit instantiations due to the `ProtoValue(google::protobuf::Value*)` constructor which calls all the other ones, but the linker didn't like that